### PR TITLE
Convert defensive sub withdraw

### PR DESCRIPTION
**Pull request by trevan**
This converts the defensive sub withdraw step.

It is very similar to the offensive sub withdraw step.  The only difference is the offensive one checks for undefended subs and prevents the retreat while the defensive one checks to see if everything has been retreated and marks the game over.  I think both of that can be reworked in separate steps but that is for a later refactoring.

## Functional Changes

[] New map or map update
[] New Feature
[] Feature update or enhancement
[] Feature Removal
[x] Code Cleanup or refactor
[] Configuration Change
[] Problem fix
[] Other:   

## Testing
Create a saved game with defensive sub retreat and verified that it loads.

## Additional Notes to Reviewer
This also unifies the logic for the step names and the executable just like with the offensive sub retreat step (#6719).  See the notes there on what was unified.

I also noticed that I had no tests for the step names in the offensive sub retreat test file.  I added those tests as part of this PR.

### CODE REVIEW

**Comment by DanVanAtta** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreat.java  7
- game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreatTest.java  1
         

Complexity decreasing per file
==============================
+ game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/BattleSteps.java  -5
         

Clones removed
==============
+ game-core/src/test/java/games/strategy/triplea/delegate/battle/MustFightBattleExecutablesTest.java  -6
+ game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/BattleStepsTest.java  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5766575&bid=18873717)

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreat.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        RetreatType.SUBS,
+        bridge,
+        battleState.getEmptyOrFriendlySeaNeighbors(
+            CollectionUtils.getMatches(battleState.getDefendingUnits(), Matches.unitCanEvade())));
+
+    if (getOrder() == SUB_DEFENSIVE_RETREAT_AFTER_BATTLE) {
+      // If no defenders left, then battle is over. The reason we test a "second" time here,
+      // is because otherwise the battle will try and do one more round and nothing will
+      // happen in that round.
+      if (battleState.getDefendingUnits().isEmpty()) {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreat.java:78]: Codacy found an issue: [These nested if statements could be combined](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5766575)

**Comment by trevan** [comment]
I also thought of creating a parent class for `OffensiveSubsRetreat` and `DefensiveSubsRetreat` to share the common code, but I don't believe that will be save game compatible.

**Comment by DanVanAtta** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreat.java  7
- game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreatTest.java  1
         

Complexity decreasing per file
==============================
+ game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/BattleSteps.java  -5
         

Clones removed
==============
+ game-core/src/test/java/games/strategy/triplea/delegate/battle/MustFightBattleExecutablesTest.java  -6
+ game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/BattleStepsTest.java  -8
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5766575&bid=18873717)

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreat.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    if (getOrder() == SUB_DEFENSIVE_RETREAT_BEFORE_BATTLE && isDestroyerPresent()) {
+      // only check for destroyers if subs can retreat before battle
+      // because the destroyer could be killed during the battle which would
+      // allow the sub to withdraw at the end of the battle
+      return List.of();
+    }
+
+    if (Properties.getSubmersibleSubs(battleState.getGameData())) {
+      return List.of(battleState.getDefender().getName() + SUBS_SUBMERGE);
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreat.java:43]: nit, any thoughts to use a ternary here to isolate the different parts of the 'if/else' block as much as possible?

EG:
```
String stepName =  battleState.getDefender().getName() + (Properties.getSubmersibleSubs(battleState.getGameData() ? SUBS_SUBMERGE : SUBS_WITHDRAW);
return List.of(stepName);
```

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreat.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    } else {
+      return List.of(battleState.getDefender().getName() + SUBS_WITHDRAW);
+    }
+  }
+
+  @Override
+  public Order getOrder() {
+    if (Properties.getSubRetreatBeforeBattle(battleState.getGameData())) {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/DefensiveSubsRetreat.java:51]: Ditto on ternary usage here, any objections or thoughts to use it?

**Pull request merged**